Exercises/Modules/Chapter 17/Soln17_05/LinkedList.cppm

```diff
@@ -0,0 +1,274 @@
+{
+	// Note: this function is strongly exception-safe because
+	// any and all change occur after the line annotated with (*),
+	// which is the only line that may throw.
+
+	Node* oldHead{ m_head };
+	m_head = new Node{ value };  // (*)
+	++m_size;
+
+	if (oldHead)
+	{
+		oldHead->m_previous = m_head;
+		m_head->m_next = oldHead;
+	}
+	else  // list was empty before, and now has one element
+	{
+		m_tail = m_head;
+	}
+}
+
+// Template for member functions that add an object to the tail of the list
+template<typename T>
+void LinkedList<T>::push_back(const T& value)
+{
+	// Note: this function is strongly exception-safe because
+	// any and all change occur after the line annotated with (*),
+	// which is the only line that may throw.
+
+	Node* oldTail{ m_tail };
+	m_tail = new Node{value};	 // (*)
+	++m_size;                 
+
+	if (oldTail)
+	{
+		oldTail->m_next = m_tail;
+		m_tail->m_previous = oldTail;
+	}
+	else  // list was empty before, and now has one element
+	{
+		m_head = m_tail;
+	}
+}
+
+// Template for member functions that remove an object from the head of the list
+template<typename T> 
+void LinkedList<T>::pop_front()
+{
+	Node* oldHead{ m_head };
+	if (oldHead == m_tail)
+	{
+		m_head = m_tail = nullptr;
+	}
+	else
+	{
+		m_head = oldHead->m_next;
+		m_head->m_previous = nullptr;
+	}
+	
+	--m_size;
+	delete oldHead;
+}
+
+// Template function member to remove an object from the tail of the list
+template<typename T>
+void LinkedList<T>::pop_back()
+{
+	Node* oldTail{ m_tail };
+	if (oldTail == m_head)
+	{
+		m_head = m_tail = nullptr;
+	}
+	else
+	{
+		m_tail = oldTail->m_previous;
+		m_tail->next = nullptr;
+	}
+
+	--m_size;
+	delete oldTail;
+}
+
+// Template function members to get the object at the head of the list
+template<typename T>
+T& LinkedList<T>::front()
+{
+	return m_head->value;
+}
+template<typename T>
+const T& LinkedList<T>::front() const
+{
+	return m_head->value;
+}
+
+// Template function members to get the object at the tail of the list
+template<typename T>
+T& LinkedList<T>::back()
+{
+	return m_tail->value;
+}
+template<typename T>
+const T& LinkedList<T>::back() const
+{
+	return m_tail->value;
+}
+
+// Check whether list is empty or not
+template<typename T>
+bool LinkedList<T>::empty() const
+{
+	return m_size == 0;  // (or m_head == nullptr, or m_tail == nullptr)
+}
+
+// Template to get the size of a list
+template<typename T> 
+size_t LinkedList<T>::size() const
+{
+	return m_size;
+}
+
+// Template to clear a list
+template<typename T>
+void LinkedList<T>::clear()
+{
+	// Use existing functions (avoid code duplication!)
+	while (!empty()) pop_front();
+}
+
+// Member function template to swap two lists
+template<typename T>
+void LinkedList<T>::swap(LinkedList& other) noexcept
+{
+	std::swap(m_head, other.m_head);
+	std::swap(m_tail, other.m_tail);
+	std::swap(m_size, other.m_size);
+}
+
+// Definition of the nested Iterator class
+template<typename T>
```

Comment:
Export?
You do it with Truckload::Iterator.
If you change it, change in all LinkedList.cppm files.

Reply:
Did not mean to do it for `Truckload::Iterator`. That's the result of a find-replace. I guess nested classes of an exported class definition do not need to be exported, but who knows. Time will tell...